File: packages/java/src/main/java/fyi/sererr/DebugInfoAdapter.java
package fyi.sererr;

import java.util.ArrayList;
import java.util.List;

/**
 * Adapter that renders a sererr capture into the {@link DebugInfo}
 * shape (wire-compatible with {@code google.rpc.DebugInfo}).
 *
 * <p>{@code stackEntries} carries one line per frame, formatted as
 * {@code "  at <function> (<file>:<line>)"}, most-recent-call-first
 * within each chain entry, separated by {@code "Caused by: <type>: <message>"}
 * header lines across the chain.
 *
 * <p>{@code detail} joins {@code "<type>: <message>"} across the chain
 * with {@code "\nCaused by: "} separators.
 *
 * <p>The chain is walked outermost-first (it's stored most-causal-first;
 * this iterates in reverse) so the outermost-caught error appears first
 * in the DebugInfo output — that's the Sentry/Python convention.
 */
public final class DebugInfoAdapter {
    private DebugInfoAdapter() {}

    /** Render a {@link CapturedError} chain into a {@link DebugInfo}. */
    public static DebugInfo toDebugInfo(List<CapturedError> chain) {
        if (chain == null || chain.isEmpty()) {
            return DebugInfo.empty();
        }
        var stackEntries = new ArrayList<String>();
        var detailParts = new ArrayList<String>();
        // chain is most-causal-first; render outermost-first.
        for (int i = chain.size() - 1, idx = 0; i >= 0; i--, idx++) {
            CapturedError entry = chain.get(i);
            String header = entry.type() + ": " + entry.message();
            if (idx == 0) {
                detailParts.add(header);
                stackEntries.add(header);
            } else {
                detailParts.add("Caused by: " + header);
                stackEntries.add("Caused by: " + header);
            }
            for (StackFrame frame : entry.frames()) {
                stackEntries.add(formatFrameLine(frame));
            }
        }
        return new DebugInfo(stackEntries, String.join("\n", detailParts));
    }

    static String formatFrameLine(StackFrame frame) {
        String location;
        if (frame.file().isEmpty()) {
            location = "";
        } else if (frame.line() > 0) {
            location = " (" + frame.file() + ":" + frame.line() + ")";
        } else {
            location = " (" + frame.file() + ")";
        }
        String fn = frame.function().isEmpty() ? "<unknown>" : frame.function();
        return "  at " + fn + location;
    }
}

File: packages/csharp/src/DebugInfoAdapter.cs
using System.Globalization;
using System.Text;

namespace Sererr;

/// <summary>
/// Renders a captured chain into the <see cref="DebugInfo"/> shape
/// (wire-compatible with <c>google.rpc.DebugInfo</c>).
///
/// Output format matches the Rust reference implementation byte-for-byte:
/// <list type="bullet">
///   <item><c>StackEntries</c>: one <c>"  at &lt;fn&gt; (&lt;file&gt;:&lt;line&gt;)"</c>
///     line per frame, most-recent-first, separated by
///     <c>"Caused by: &lt;type&gt;: &lt;message&gt;"</c> headers.</item>
///   <item><c>Detail</c>: joined <c>"&lt;type&gt;: &lt;message&gt;"</c>
///     across the chain with <c>"\nCaused by: "</c> separators.</item>
/// </list>
/// </summary>
public static class DebugInfoAdapter
{
    /// <summary>Build a <see cref="DebugInfo"/> from a captured chain.</summary>
    public static DebugInfo ToDebugInfo(IReadOnlyList<CapturedError> chain)
    {
        if (chain.Count == 0)
        {
            return new DebugInfo();
        }

        var stackEntries = new List<string>();
        var detailParts = new List<string>();

        // chain is stored most-causal-first. The user-facing rendering
        // is most-recent-first — walk in reverse.
        for (var revIdx = 0; revIdx < chain.Count; revIdx++)
        {
            var entry = chain[chain.Count - 1 - revIdx];
            var header = $"{entry.Type}: {entry.Message}";
            detailParts.Add(revIdx == 0 ? header : $"Caused by: {header}");

            stackEntries.Add(revIdx == 0 ? header : $"Caused by: {header}");

            foreach (var frame in entry.Frames)
            {
                stackEntries.Add(FormatFrameLine(frame));
            }
        }

        return new DebugInfo
        {
            StackEntries = stackEntries,
            Detail = string.Join("\n", detailParts),
        };
    }

    private static string FormatFrameLine(StackFrame frame)
    {
        string location;
        if (string.IsNullOrEmpty(frame.File))
        {
            location = "";
        }
        else if (frame.Line > 0)
        {
            location = $" ({frame.File}:{frame.Line.ToString(CultureInfo.InvariantCulture)})";
        }
        else
        {
            location = $" ({frame.File})";
        }

        var function = string.IsNullOrEmpty(frame.Function) ? "<unknown>" : frame.Function;
        return $"  at {function}{location}";
    }
}
